token/program-2022/src/extension/mod.rs

```diff
@@ -511,31 +498,24 @@ impl<'data, S: BaseState> StateWithExtensionsMut<'data, S> {
     fn get_first_extension_type(&self) -> Result<Option<ExtensionType>, ProgramError> {
         get_first_extension_type(self.tlv_data)
     }
+}
 
-    /// Compares the length of an extension with the currently used TLV buffer to determine if
-    /// reallocation is needed. If so, returns Some(v), where v is the difference between current
-    /// space and needed.
-    #[allow(dead_code)]
-    pub(crate) fn realloc_needed(
-        &self,
-        new_extension: ExtensionType,
-    ) -> Result<Option<usize>, ProgramError> {
-        let mut extensions = self.get_extension_types()?;
-        if !extensions.contains(&new_extension) {
-            extensions.push(new_extension);
-        }
-        let new_needed_tlv_len = ExtensionType::get_total_tlv_len(&extensions);
-        if self.tlv_data.len() >= new_needed_tlv_len {
-            Ok(None)
-        } else {
-            let mut diff = new_needed_tlv_len - self.tlv_data.len(); // arithmetic safe because of if clause
-            if self.account_type.is_empty() {
-                diff = diff
-                    .saturating_add(size_of::<AccountType>())
-                    .saturating_add(BASE_ACCOUNT_LENGTH.saturating_sub(S::LEN));
-            }
-            Ok(Some(diff))
+/// If AccountType is uninitialized, set it to the BaseState's ACCOUNT_TYPE;
+/// if AccountType is already set, check is set correctly for BaseState
+pub fn set_account_type<S: BaseState>(input: &mut [u8]) -> Result<(), ProgramError> {
```

Comment:
Since we only need to consider this for accounts at the moment, we don't need to worry about this case, but I want to get it out there.

In a vacuum, if you pass in a buffer that currently contains just a base mint and call this with `S = Account`, it will succeed, correct? `type_and_tlv_indices` only checks for the opposite case of passing in a buffer containing a base `Account` but call it with `S = Mint`.  I wonder if just to be totally totally safe, this should check the `is_initialized` byte to make sure it's actually an `Account` if `S = Account`.  Doesn't make it too generic though...

Probably doesn't need to get done here, and probably won't become an issue, but this exact case has been in the back of my mind for awhile now.

Reply:
Mm, yeah good point. I originally had `let base = S::unpack(base_data)?;` in there, which would protected against that, but deleted for expense reasons. I'm going to merge this as-is, but will file an issue for us to fix in a follow-up.